### PR TITLE
[cisco_asa] Fix Event code 106023 - Source/Destination IP not being parsed into respective source.ip or destination.ip field when interface nameif has a full colon (:)

### PR DESCRIPTION
**Pull request by aleksmaus**
[cisco_asa] Fix Event code 106023 - Source/Destination IP not being parsed into respective source.ip or destination.ip field when interface nameif has a full colon (:)

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Closes https://github.com/elastic/integrations/issues/9184

### CODE REVIEW

**Comment by aleksmaus** [comment]
hmm, actually since this is completely might the user defined, hold on a sec.... will refined the regex

File: packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-asa-106023-fix.log [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Apr 17 2020 14:00:31 SNL-ASA-VPN-A01 : %ASA-4-106023: Deny icmp src Inside:10.123.123.123 dst Outside:10.123.123.123 (type 11, code 0) by access-group "Inside_access_in" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst v2:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
```
Comment (pkoutsovasilis) [inline packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-asa-106023-fix.log:3]: can you add also an example with ipv6 ips which are full of colons? just to make sure that nothing breaks in the former case

Comment (aleksmaus) [reply packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-asa-106023-fix.log:3]: there is already one, search
```
Apr 27 2020 02:03:03 dev01: %ASA-4-106023: Deny icmp src OUTSIDE:2a02:cf40:add:4002:91f2:a9b2:e09a:6fc6 dst OUTSIDE:2a02:cf40:add:4002:91f2:a9b2:e09a:6fc6 (type 128, code 0) by access-group "OUTSIDE_in"
```

Comment (pkoutsovasilis) [reply packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-asa-106023-fix.log:3]: yes one already with an iface without colon it exists, but you are fixing support for ifaces with colons in them right?! So can we add one that handles this case?

**Comment by aleksmaus** [comment]
After doing some digging around, it doesn't look like we have a robust support for parsing ipv6 case if the hostname has prefix that contains ":".
Problem is recognizing the ":" for IPv6 address from the  possible ":" in the the hostname. At the moment we don't have immediate cases where hostname contains ":" in the name with IPv6. So will attempt to cover the immediate case with IPv4 and then will see if IPv6 case ever materializes.

Currently having an issue with Grok. The expression works fine in the "Grok" debugger.
Sample Data:
```
Deny udp src outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
```

Grok pattern:
```
^Deny ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{IFACENAME:_temp_.cisco.source_interface}:%{IPV4ORHOST:source.address}(/%{NUMBER:source.port})?\s*(\(%{CISCO_USER_OR_SGT_SRC}\) )?dst %{IFACENAME:_temp_.cisco.destination_interface}:%{IPV4ORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group "%{NOTSPACE:_temp_.cisco.list_id}"
```

Custom Patterns:
```
HOSTNAME \b(?:[0-9A-Za-z][0-9A-Za-z\-_]{0,62})(?:\.(?:[0-9A-Za-z][0-9A-Za-z\-_]{0,62}))*(\.?|\b)
IPORHOST (?:%{IP}|%{HOSTNAME})
IPV4ORHOST (?:%{IPV4}|%{HOSTNAME})
NOTCOLON [^:]*
IFACENAME (.*:)?[^:]*
USERNAME [a-zA-Z0-9._'-]+
CISCO_USER_OR_SGT_SRC (?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt}|\((?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt})\))
CISCO_USER (?:\*\*\*\*\*|(?:(?:LOCAL\\)?(?:%{HOSTNAME}\\)?%{USERNAME}\$?(?:@%{HOSTNAME})?%{CISCO_SGT}?)|[^$]+)
CISCO_SGT (?:, *)?%{NUMBER}(?::%{WORD})?
```

Returns:
```
{
  "destination": {
    "port": "53",
    "address": "172.16.1.3"
  },
  "_temp_": {
    "cisco": {
      "destination_interface": "v3:inside",
      "list_id": "outside_acl",
      "source_interface": "outside"
    }
  },
  "source": {
    "address": "10.8.1.9"
  },
  "network": {
    "transport": "udp"
  }
}
```

But the same thing doesn't work in the ingest pipeline, cliping the destination_interface:
```
            "cisco": {
                "asa": {
                    "destination_interface": "v3",
                    "rule_name": "outside_acl",
                    "source_interface": "outside"
                }
            },
```

It's interesting, if I add prefixes to both ```src``` and ```dst``` or only ```src``` the pipeline works fine. 
It's not helpful that the same processor with ingest pipeline returns different result than the grok debugger output. 

Will get back to this later.

**Comment by aleksmaus** [comment]
KK. Figured out the problem with my previous attempt. The original regex 
```
^Deny ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{NOTCOLON:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\s*(\(%{CISCO_USER_OR_SGT_SRC}\) )?dst %{NOTCOLON:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group "%{NOTSPACE:_temp_.cisco.list_id}"
```

actually matches when the ```dst``` has prefix:
```
Deny udp src outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group \"outside_acl\" [0x0, 0x0]
```
and doesn't match any other permutation of src or dst prefixes

That's why I saw the different result in the pipeline simulator. 
Example playground

```
POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    "description": "_description",
    "processors": [
      {
        "grok": {
          "field": "message",
          "patterns": [
            "^Deny ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{NOTCOLON:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\\s*(\\(%{CISCO_USER_OR_SGT_SRC}\\) )?dst %{NOTCOLON:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group \"%{NOTSPACE:_temp_.cisco.list_id}\""         ],
          "pattern_definitions": {
            "HOSTNAME": "\\b(?:[0-9A-Za-z][0-9A-Za-z\\-_]{0,62})(?:\\.(?:[0-9A-Za-z][0-9A-Za-z\\-_]{0,62}))*(\\.?|\\b)",
            "IPORHOST": "(?:%{IP}|%{HOSTNAME})",
            "NOTCOLON": "[^:]*",
            "IFACENAME": "(.*)?%{NOTCOLON}",
            "USERNAME": "[a-zA-Z0-9._'-]+",
            "CISCO_USER_OR_SGT_SRC": "(?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt}|\\((?:%{CISCO_USER:_temp_.cisco.source_user_or_sgt}|%{CISCO_SGT:_temp_.cisco.source_user_or_sgt})\\))",
            "CISCO_USER": "(?:\\*\\*\\*\\*\\*|(?:(?:LOCAL\\\\)?(?:%{HOSTNAME}\\\\)?%{USERNAME}\\$?(?:@%{HOSTNAME})?%{CISCO_SGT}?)|[^$]+)",
            "CISCO_SGT": "(?:, *)?%{NUMBER}(?::%{WORD})?"
          }
        }
      }
    ]
  },
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "message0": "Deny udp src outside:10.8.1.9 dst inside:172.16.1.3/53 by access-group \"outside_acl\" [0x0, 0x0]",
        "message1": "Deny udp src v2:outside:10.8.1.9 dst inside:172.16.1.3/53 by access-group \"outside_acl\" [0x0, 0x0]",
        "message": "Deny udp src outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group \"outside_acl\" [0x0, 0x0]",
        "message3": "Deny udp src v2:outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group \"outside_acl\" [0x0, 0x0]"
      }
    }
  ]
}
```

Output:
```
{
  "docs": [
    {
      "doc": {
        "_index": "index",
        "_version": "-3",
        "_id": "id",
        "_source": {
          "_temp_": {
            "cisco": {
              "destination_interface": "v3",
              "list_id": "outside_acl",
              "source_interface": "outside"
            }
          },
          "destination": {
            "address": "inside"
          },
          "message3": """Deny udp src v2:outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]""",
          "source": {
            "address": "10.8.1.9"
          },
          "message0": """Deny udp src outside:10.8.1.9 dst inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]""",
          "message": """Deny udp src outside:10.8.1.9 dst v3:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]""",
          "message1": """Deny udp src v2:outside:10.8.1.9 dst inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]""",
          "network": {
            "transport": "udp"
          }
        },
        "_ingest": {
          "timestamp": "2024-08-30T01:20:09.769757584Z"
        }
      }
    }
  ]
}
```

Will update PR shortly

**Comment by aleksmaus** [comment]
Hold on on reviewing this one. I'm not too happy that there is an additional fairly complex grok/regex for that special case, that now is checked for match first. I'll see if there is anything that can be done.

**Comment by aleksmaus** [comment]
Refined the existing regex to match better the existing data as well as the new logs when interface name contains `:`

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -314,11 +314,10 @@ processors:
       field: "message"
       description: "106023"
       patterns:
-        - ^Deny ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{NOTCOLON:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\s*(\(%{CISCO_USER_OR_SGT_SRC}\) )?dst %{NOTCOLON:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?%{DATA}by access-group "%{NOTSPACE:_temp_.cisco.list_id}"
+        - "^Deny ((protocol %{POSINT:network.iana_number})|%{NOTSPACE:network.transport}) src %{DATA:_temp_.cisco.source_interface}:%{IPORHOST:source.address}(/%{NUMBER:source.port})?\\s*(\\(%{CISCO_USER_OR_SGT_SRC}\\) )?dst %{DATA:_temp_.cisco.destination_interface}:%{IPORHOST:destination.address}(/%{NUMBER:destination.port})?[(\\s]+%{DATA}by access-group \"%{NOTSPACE:_temp_.cisco.list_id}\""
```
[the remark this answers is not in the document]
Comment (aleksmaus) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:317]: yeah, looking at the current logs there could be a space or the '('. The previous approach with using DATA for the whole thing was failing to correctly match the host/ip/port before that

File: packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-106023-iface-with-prefix.log [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src outside:10.8.1.9/54864 dst v2:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst v2:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9 dst inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src outside:10.8.1.9 dst v2:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9 dst v2:inside:172.16.1.3/53 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst inside:172.16.1.3 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src outside:10.8.1.9/54864 dst v2:inside:172.16.1.3 by access-group "outside_acl" [0x0, 0x0]
+LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst v2:inside:172.16.1.3 by access-group "outside_acl" [0x0, 0x0]
```
Comment (mjwolf) [inline packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-106023-iface-with-prefix.log:9]:
```suggestion
LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src v1:outside:10.8.1.9/54864 dst v2:inside:172.16.1.3 by access-group "outside_acl" [0x0, 0x0]
LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src outside:::abdf/548 dst v2:inside:172.16.1.3 by access-group "outside_acl" [0x0, 0x0]
LOCAL4.WARNING: fw-1 %ASA-4-106023: Deny udp src outside:10.8.1.9/54864 dst v2:inside:::2063 by access-group "outside_acl" [0x0, 0x0]
```

These are some interesting edge cases that might be good to cover in the tests. They're already working properly

Comment (aleksmaus) [reply packages/cisco_asa/data_stream/log/_dev/test/pipeline/test-106023-iface-with-prefix.log:9]: it passes the pipeline fine but the tests fail because:
```
[0] parsing field value failed: the IP "::2063" is not one of the allowed test IPs (see: https://github.com/elastic/elastic-package/blob/main/internal/fields/_static/allowed_geo_ips.txt)
[1] parsing field value failed: the IP "::abdf" is not one of the allowed test IPs (see: https://github.com/elastic/elastic-package/blob/main/internal/fields/_static/allowed_geo_ips.txt)
```

File: packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (aleksmaus) [reply packages/cisco_asa/data_stream/log/elasticsearch/ingest_pipeline/default.yml:317]: image

**Review by gogochan — APPROVED** [review]

**Review by mjwolf — APPROVED** [review]
